### ISSUE TITLE
Keyboard shortcut for "Run Function Definition"

### ISSUE THREAD

**Issue by kwoolner** [REditorSupport/vscode-R#921]
I've been working on migrating from developing in R using Eclipse and the StatET plugin (packaged as Architect at www.getarchitect.io) to VSCode. One of the little niceties that I haven't found a counterpart for so far is a keyboard shortcut that allows you run the function definition of whatever function code block the cursor is in.  

**Describe the solution you'd like**
When your cursor is in an editor window, and inside a block of R code that defines a function (e.g. myFunction <- function(x) {}), have a keyboard shortcut (in StatET it is CTRL-R, CTRL-F) that automatically detects the code that spans the function definition and submits it to the active console. The cursor remains in the editor at the same spot, and the editing window remains the active pane. For functions that are nested in other functions, the outermost function definition is run.

**Describe alternatives you've considered**
This is inspired by functionality that exists in Eclipse/StatET. The closest I have found in VSCode/R is to select the entire function definition manually and typing CTRL-ENTER to submit the code to the active console. 

**Additional context**
A brief mention of the function in Eclipse/StatET can be found at https://www.r-bloggers.com/2009/10/tips-for-using-statet-and-eclipse-for-data-analysis-in-r/ in the section Options for Running R Code

**Comment by renkun-ken** [comment]
You might want command "R: Run Selection/Line (Retain Cursor)" (`r.runSelectionRetainCursor`) which detects the code block if no selection range is specified and does retain the cursor position.

**Comment by kwoolner** [comment]
Thank you for the suggestion. I tried that out, and while it's close that doesn't seem to be doing quite what I requested. If I have a function definition like this:

```
plotme <- function(x, y, color=NULL) {
    plot(x, y, col=if (is.null(color))"black" else color)
}
```

"R: Run Selection/Line (Retain Cursor)"  works only if I am on the very first line or the very last line of code, but not in the interior line (containing the `plot()` call). It only seems to work if I'm already on a line that has a block delimiter character like `{` or `}`

Using the example from #325 

```
m <- local({
  x <- rnorm(100)
  y <- rnorm(100)
})
```

The block gets submitted if I'm on line 1 or 4,  but not 2 or 3. So ideally if my cursor was on the `x <- rnorm(100)` line, then upon submitting the version of the command I have in mind, VSCode would detect the entire (and outermost) R block containing that line, and submit that.

My use case would be that I'm modifying the interior of a (potentially long) function, and want to resend the whole function definition from an interior line, perhaps to do a quick manual test.  In the IDE I'm coming from, this functionality applied specifically to function definitions rather than arbitrary blocks of R code, probably to avoid the chance that you execute code when you think you're just changing a function definition. But either behavior could be valuable (or configurable?). I found this very handy when working with a script containing several functions for modularity. The key difference from the Run Selection/Line (Retain Cursor) command is that it would work on interior lines within a block.

Would such a command be possible? Thanks again.

**Comment by renkun-ken** [comment]
Then what's the expected behavior of the following:

```r
foo <- function(x, y) {
  bar <- function(a, b) {
    a + b
  }
  bar(x + y, x - y)
}
```

Modifying the function body of the inner function `bar` at line `a + b`. Do you always want the top-level function definition to be sent to terminal?

**Comment by kwoolner** [comment]
Yes, always the top-level. Even if I was on the `a + b` line, I would not want `bar` in the global workspace when it is intended to be a sub function of `foo`, so I would want the entire definition of `foo` resent. If I wanted to override the behavior implied by the nesting and put `bar` in the global environment, I could always select just the lines of `bar` and submit those, although that should be rare by design.

**Comment by renkun-ken** [comment]
Then we could implement a new command that detects the top-level expression range, sends the expression to terminal, and retains the cursor position.

**Comment by kwoolner** [comment]
That would be great! 

Are you thinking that such a command would be specific to function definitions, or would work with any arbitrary block of code? I'm not sure whether the "protection" of only allowing function definitions is better or worse than enabling it for all blocks of code. Could see it going either way, and either approach would work for my use case.

**Comment by renkun-ken** [comment]
Working with an arbitrary block of code would be fine, since there could be many ways to *define a function*.

**Comment by kwoolner** [comment]
ok, sounds good! Thank you

**Comment by renkun-ken** [comment]
Are you interested in implementing this feature? Do you have a way in mind to detect the top-level expression without parsing the entire document?

**Comment by kwoolner** [comment]
While not unwilling to help, it's not really in my current skillset. I'm still new to VSCode just as a user, and I don't know the first thing about implementing VSCode extensions. 

I think StatET is written in Java, which I also do not know. Not sure how Eclipse/StatET does the detection under the hood, but it works even if the Windows front end is connecting to a remote client (say R running on Linux) through an SSH tunnel, without having to have R installed on the front end. There's an "rJava" package that facilitates communication between the ends. Not sure if they are relying on using the R session itself to aid in parsing, or if the detection is being done solely by the front end client.

If I had to speculate, I'd guess that they are looking for the `function` R keyword in the doc, and maybe just finding the enclosing block around each instance? Then if the cursor is on a line number within a known block, it knows what range to send?

**Comment by kwoolner** [comment]
FWIW, I just tested Eclipse/StatET to verify that it handled nesting the way I remembered. It did for at least 5 levels deep (after which I stopped). In each case, it submitted the very top-most enclosing function definition.

I also spent a bit of time perusing the Java source, but couldn't follow it enough to really know what they are doing. But I was able to determine that they do detect `function` in the code being edited as a token to keep track of (along with many other R keywords). But how it specifically determines what the topmost expression for a given line of code is, and whether it is a function definition, isn't clear.
